proof/crefine/RISCV64/Machine_C.thy

```diff
@@ -130,6 +130,528 @@ lemma dmo_if:
+(* Count leading and trailing zeros. *)
+
+(* FIXME BV: move *)
+lemma word_ctz_max:
+  "word_ctz w \<le> size w"
+  unfolding word_ctz_def
+  by (rule order_trans[OF List.length_takeWhile_le], clarsimp simp: word_size)
+
+(* FIXME BV: move *)
+lemma scast_of_nat_small:
+  "x < 2 ^ (LENGTH('a) - 1) \<Longrightarrow> scast (of_nat x :: 'a :: len word) = (of_nat x :: 'b :: len word)"
+  apply (rule sym, subst word_unat.inverse_norm)
+  apply (simp add: scast_def word_of_int[symmetric]
+                   of_nat_nat[symmetric] unat_def[symmetric])
+  apply (simp add: int_eq_sint unat_of_nat)
+  done
+
+(* FIXME BV: move *)
+lemmas casts_of_nat_small = ucast_of_nat_small scast_of_nat_small
+
+(* FIXME BV: move *)
+lemma hoarep_Seq_nothrow:
+  assumes c: "hoarep \<Gamma> \<Theta> F P c Q {}"
+  assumes d: "hoarep \<Gamma> \<Theta> F Q d R A"
+  shows "hoarep \<Gamma> \<Theta> F P (Seq c d) R A"
+  by (rule hoarep.Seq[OF HoarePartialDef.conseqPrePost[OF c] d]; simp)
+
+definition clz32_step where
+  "clz32_step i \<equiv> \<acute>mask___unsigned :== \<acute>mask___unsigned >> unat ((1::32 sword) << unat i);;
+                 \<acute>bits :== SCAST(32 signed \<rightarrow> 32) (if \<acute>mask___unsigned < \<acute>x___unsigned then 1 else 0) << unat i;;
+                 Guard ShiftError \<lbrace>\<acute>bits < SCAST(32 signed \<rightarrow> 32) 0x20\<rbrace> (\<acute>x___unsigned :== \<acute>x___unsigned >> unat \<acute>bits);;
+                 \<acute>count :== \<acute>count - \<acute>bits"
+
+definition clz32_invariant where
+  "clz32_invariant i s \<equiv> {s'.
+   mask___unsigned_' s' \<ge> x___unsigned_' s'
+   \<and> of_nat (word_clz (x___unsigned_' s')) + count_' s' = of_nat (word_clz (x___unsigned_' s)) + 32
+   \<and> mask___unsigned_' s' = mask (2 ^ unat i)}"
+
+\<comment>\<open> This function returns an expression stating that "if xs contains an element satisfying P, then
+    the nth element of xs is the *first* that satisfies P". \<close>
+definition first_in_list where
+  "first_in_list P n xs \<equiv> (\<forall>i. i < n \<longrightarrow> i < length xs \<longrightarrow> \<not> P (xs ! i)) \<and>
+                           (n < length xs \<longrightarrow> P (xs ! n))"
+
+\<comment>\<open> There is always some n at which P (x ! n) or P contains no such element \<close>
+lemma first_in_list_exists:
+  "\<exists>n. first_in_list P n xs"
+  apply (induction xs; simp add: first_in_list_def)
+  apply (erule exE)
+  apply (case_tac "P x1")
+   apply (rule_tac x=0 in exI, simp)
+  apply (rule_tac x="Suc n" in exI, clarsimp)
+  apply (case_tac i; simp)
+  done
+
+\<comment>\<open> If you know where P first fails to be true, then you know how to compute takeWhile with take. \<close>
+lemma takeWhile_take_equiv:
+  "first_in_list (not P) n xs \<Longrightarrow> takeWhile P xs = take n xs"
+  apply (rule takeWhile_eq_take_P_nth)
+  apply (clarsimp simp: first_in_list_def pred_neg_def)
+  apply (clarsimp simp: first_in_list_def pred_neg_def)
+  done
```

Comment:
I was mildly dubious about whether this deserved a definition, but I think you've convinced me.

If you feel like it, you could tune this slightly:
- Consider inverting the condition, since it gives nicer theorems about `takeWhile` and `dropWhile`, and you only ever use it inverted anyway.
- With the inverted condition, it needs a new name. TBH, `first_in_list` didn't strike me as a great name. Perhaps `list_while_len`? Still not amazing, but it attempts to capture that something is true up to some length.
- Rephrase `takeWhile_take_equiv` as an equivalence:
  ```
  lemma list_while_len_takeWhile:
    "list_while_len P n xs == takeWhile P xs = take n xs"
  ```
  There'd be a similar rule for `dropWhile` and `drop`, but no need to add that if you don't need it.

Reply:
Done.